packages/dashboard/src/components/queryEditor/iotSiteWiseQueryEditor/queryExtender/queryExtender.ts

```diff
@@ -0,0 +1,99 @@
+import type { ModeledDataStream } from '../modeledDataStreamQueryEditor/modeledDataStreamExplorer/types';
+import type { UnmodeledDataStream } from '../unmodeledDataStreamExplorer/types';
+import type { IoTSiteWiseDataStreamQuery } from '~/types';
+
+type Query = IoTSiteWiseDataStreamQuery;
+
+export class QueryExtender {
+  readonly #currentQuery: Query;
+
+  constructor(currentQuery: Query = {}) {
+    this.#currentQuery = currentQuery;
+  }
+
+  public extendAssetQueries(modeledDataStreams: ModeledDataStream[]): Query {
+    const currentAssetQueries = this.#currentQuery.assets ?? [];
+    const newAssetQueries = this.#createAssetQueries(modeledDataStreams);
+    const dedupedAssetQueries = this.#dedupeAssetQueries([...currentAssetQueries, ...newAssetQueries]);
+    const extendedQuery = {
+      ...this.#currentQuery,
+      assets: dedupedAssetQueries,
+    };
+
+    return extendedQuery;
+  }
+
+  #createAssetQueries(modeledDataStreams: ModeledDataStream[]): NonNullable<Query['assets']> {
+    const assetQueriesWithProperties = modeledDataStreams
+      .filter(this.#isModeledDataStream)
+      .map<NonNullable<Query['assets']>[number]>(({ assetId, propertyId: propertyId }) => ({
+        assetId,
+        properties: [{ propertyId }],
+      }));
+
+    return assetQueriesWithProperties;
+  }
+
+  #isModeledDataStream(dataStream: ModeledDataStream | UnmodeledDataStream): dataStream is ModeledDataStream {
```

Comment:
arrow functions are better ? so that you do not need to bind to this explicitly

Reply:
I default to using method declarations instead of arrow functions in classes as method declarations are shared across class instances and arrow functions are not. This can cause performance problems at scale, as every class instance creates its own instance of the methods.

This rationale doesn't apply to this class, as there are not many instances being created, but I would rather not think about it, so I just default to method declarations.

I still use arrow functions in classes to handle to avoid the binding problem. It's rare I have a binding issue that needs to be handled. If I was using the method as a callback, I might reach for an arrow function, especially if I am handing control to the browser with an event listener, setTimeout, or other browser API.

Reply:
It's also expensive to instantiate a class with multiple arrow functions. When you call `new`, you create all the functions on the spot. What's the point of a class if you don't utilize its ability to act as a template? 😛